File: java/leetcode/src/main/java/com/leetcode/list/Insertion_Sort_List_147.java
package com.leetcode.list;

public class Insertion_Sort_List_147 {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode();
        ListNode cur = head;
        while (cur != null) {
            ListNode nextCur = cur.next;
            insert(dummy, cur);
            cur = nextCur;
        }
        return dummy.next;
    }

    void insert(ListNode dummy, ListNode node) {
        ListNode pre = dummy;
        ListNode cur = dummy.next;
        while(cur != null) {
            if(node.val > cur.val) {
                pre = cur;
                cur = cur.next;
            }else {
                pre.next = node;
                node.next = cur;
                break;
            }
        }
        if (cur == null) {
            pre.next = node;
            node.next = null;
        }
    }
}

File: js/algrithm/list/Insertion Sort List_147.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 
var insertionSortList = function(head) {
    const dummy = new ListNode()
    let cur = head
    while (cur) {
        let next = cur.next
        cur.next = null
        insert(dummy, cur)
        cur = next
    }
    return dummy.next
};

function insert(dummy, one) {
    let pre = dummy
    let cur = dummy.next
    while (cur && cur.val < one.val) {
        pre = cur
        cur = cur.next
    }
    pre.next = one
    one.next = cur
}